pkg/controller/azure/compute/akscluster_test.go

```diff
@@ -268,6 +288,146 @@ func TestReconcile(t *testing.T) {
+		}, nil
+	}
+	mockAKSSetupClient.MockCreateServicePrincipal = func(ctx context.Context, spID, appID string) (*graphrbac.ServicePrincipal, error) {
+		return &graphrbac.ServicePrincipal{
+			ObjectID: to.StringPtr("da804153-3faa-4c73-9fcb-0961387a31f9"),
+		}, nil
+	}
+	mockAKSSetupClient.MockCreateRoleAssignment = func(ctx context.Context, sp, vnetSubnetID, name string) (result *authorization.RoleAssignment, err error) {
+		return &authorization.RoleAssignment{}, nil
+	}
+	mockAKSSetupClient.MockCreateOrUpdateBegin = func(ctx context.Context, instance computev1alpha1.AKSCluster, clusterName, appID, spSecret string) ([]byte, error) {
+		return []byte("mocked marshalled create future"), nil
+	}
+	mockAKSSetupClient.MockGet = func(ctx context.Context, instance computev1alpha1.AKSCluster) (containerservice.ManagedCluster, error) {
+		return containerservice.ManagedCluster{
+			ID: to.StringPtr("fcb4e97a-c3ea-4466-9b02-e728d8e6764f"),
+			ManagedClusterProperties: &containerservice.ManagedClusterProperties{
+				ProvisioningState: to.StringPtr("Succeeded"),
+				Fqdn:              to.StringPtr("crossplane-aks.foo.azure.com"),
+			},
+		}, nil
+	}
+	mockAKSSetupClient.MockListClusterAdminCredentials = func(ctx context.Context, instance computev1alpha1.AKSCluster) (containerservice.CredentialResults, error) {
+		return containerservice.CredentialResults{
+			Kubeconfigs: &[]containerservice.CredentialResult{{Value: &kubecfg}},
+		}, nil
+	}
+
+	// Setup the Manager and Controller.  Wrap the Controller Reconcile function so it writes each request to a
+	// channel when it is finished.
+	mgr, err := manager.New(cfg, manager.Options{MetricsBindAddress: ":8081"})
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	c := mgr.GetClient()
+
+	r := NewAKSClusterReconciler(mgr, mockAKSSetupClientFactory, clientset)
+	recFn, requests := SetupTestReconcile(r)
+	controller := &AKSClusterController{
+		Reconciler: recFn,
+	}
+	g.Expect(controller.SetupWithManager(mgr)).NotTo(gomega.HaveOccurred())
+	defer close(StartTestManager(mgr, g))
+
+	// create the provider object and defer its cleanup
+	provider := testProvider(testSecret([]byte("testdata")))
+	err = c.Create(ctx, provider)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	defer c.Delete(ctx, provider)
+
+	// Create the AKS cluster object and defer its clean up
+	instance := testInstanceInSubnet(provider)
+	err = c.Create(ctx, instance)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	defer cleanupAKSCluster(t, g, c, requests, instance)
+
+	// first reconcile loop should start the create operation
+	g.Eventually(requests, timeout).Should(gomega.Receive(gomega.Equal(expectedRequest)))
+
+	// after the first reconcile, the create operation should be saved on the running operation field,
+	// and the following should be set:
+	// 1) cluster name
+	// 2) application object ID
+	// 3) service principal ID
+	// 4) "creating" condition
+	expectedStatus := computev1alpha1.AKSClusterStatus{
```

Comment:
this unit test pattern is definitely not ideal, it'll be nice to upgrade AKS to the managed reconciler pattern too eventually.